selfservice/strategy/webauthn/login.go

```diff
@@ -72,27 +73,62 @@ func (s *Strategy) PopulateLoginMethod(r *http.Request, requestedAAL identity.Au
 		if identifier == "" {
 			return nil
 		}
 
-		if err := s.populateLoginMethod(r, sr, id, text.NewInfoSelfServiceLoginWebAuthn(), ""); errors.Is(err, ErrNoCredentials) {
+		if err := s.populateLoginMethod(r, loginFlow, id, text.NewInfoSelfServiceLoginWebAuthn(), ""); errors.Is(err, ErrNoCredentials) {
 			return nil
 		} else if err != nil {
 			return err
 		}
 
-		sr.UI.SetCSRF(s.d.GenerateCSRFToken(r))
-		sr.UI.SetNode(node.NewInputField("identifier", identifier, node.DefaultGroup, node.InputAttributeTypeHidden))
+		loginFlow.UI.SetCSRF(s.d.GenerateCSRFToken(r))
+		loginFlow.UI.SetNode(node.NewInputField("identifier", identifier, node.DefaultGroup, node.InputAttributeTypeHidden))
 		return nil
 	}
 
-	sr.UI.SetCSRF(s.d.GenerateCSRFToken(r))
-	sr.UI.SetNode(node.NewInputField("identifier", "", node.DefaultGroup, node.InputAttributeTypeText, node.WithRequiredInputAttribute).WithMetaLabel(text.NewInfoNodeLabelID()))
-	sr.UI.GetNodes().Append(node.NewInputField("method", "webauthn", node.WebAuthnGroup, node.InputAttributeTypeSubmit).WithMetaLabel(text.NewInfoSelfServiceLoginWebAuthn()))
+	if s.d.Config().WebAuthnEnableDiscoverableCredentials(r.Context()) {
+		web, err := webauthn.New(s.d.Config().WebAuthnConfig(r.Context()))
+		if err != nil {
+			return errors.WithStack(herodot.ErrInternalServerError.WithReasonf("Unable to initiate WebAuthn.").WithDebug(err.Error()))
+		}
+
+		ca, sessionData, err := web.BeginDiscoverableLogin()
+		if err != nil {
+			return errors.WithStack(herodot.ErrInternalServerError.WithReasonf("Unable to initiate discoverable WebAuthn login.").WithDebug(err.Error()))
+		}
+
+		loginFlow.InternalContext, err = sjson.SetBytes(loginFlow.InternalContext, flow.PrefixInternalContextKey(s.ID(), InternalContextKeySessionData), sessionData)
+		if err != nil {
+			return errors.WithStack(herodot.ErrInternalServerError.WithReasonf("Unable to update internal context.").WithDebug(err.Error()))
+		}
+		ca.Response.AllowedCredentials = []protocol.CredentialDescriptor{}
+		injectWebAuthnOptions, err := json.Marshal(ca)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		loginFlow.UI.SetCSRF(s.d.GenerateCSRFToken(r))
+		loginFlow.UI.SetNode(NewWebAuthnScript(urlx.AppendPaths(s.d.Config().SelfPublicURL(r.Context()), webAuthnRoute).String(), jsOnLoad))
+		loginFlow.UI.SetNode(node.NewInputField("webauthn_options", string(injectWebAuthnOptions), node.WebAuthnGroup, node.InputAttributeTypeHidden))
+		loginFlow.UI.SetNode(NewWebAuthnLoginTrigger(string(injectWebAuthnOptions)).WithMetaLabel(text.NewInfoSelfServiceLoginWebAuthn()))
+		loginFlow.UI.Nodes.Upsert(NewWebAuthnLoginInput())
+		loginFlow.UI.Nodes.Upsert(node.NewInputField("discovered", "true", node.WebAuthnGroup, node.InputAttributeTypeHidden))
+
+		identifierNode := loginFlow.UI.Nodes.Find("identifier")
+		if identifierNode != nil {
+			if a, ok := identifierNode.Attributes.(*node.InputAttributes); ok {
+				a.Autocomplete = a.Autocomplete + " webauthn"
+			}
+		}
```

Comment:
This likely needs a better solution.

We want to add `"webauthn"` to any identifier field shown on the login page, so the browser can show the autofil UI. But with this change, the webauthn strategy no longer renders an `id` input field.